### PR TITLE
add elasticsearch connector

### PR DESCRIPTION
**Pull request by harbby**
# presto-elasticsearch
presto connector for elasticsearch

## install
es 2.x
```
connector.name=elasticsearch

elasticsearch.cluster.name=ideal-bigdata
elasticsearch.transport.hosts=localhost:9300
```

es 5.x
```
connector.name=elasticsearch5

elasticsearch.cluster.name=ideal-bigdata
elasticsearch.transport.hosts=localhost:9300
```

es 6.x
```
connector.name=elasticsearch6

elasticsearch.cluster.name=ideal-bigdata
elasticsearch.transport.hosts=localhost:9300
```
look: Cannot be used simultaneously

## SQL Usage

```SELECT * FROM test1 WHERE age >30 AND city = 'world'```

```desc test1```

## Beyond SQL

* Search

select * from test1 where  _name = match_query('Lucy')

* Search match_phrase

select * from test1 where  _name = match_phrase('Lucy')

* Strong pushDown

select * from test1 where _dsl = '{"query":{"match":{"city":{"query":"world"}}}}' and _name = match_query('Lucy')

### pushDown
+ _ field is an extension field
Use the _ field and it will be pushed down to es for search

+ -dsl field will be fully pushed down for custom dsl

## session parameter

| Name                                    | Default        | Type    | Description
| -------------------------------------   | -------------- | ------- | ----------------------------------------------------------------------------------
| es6.optimize_split_shards_enabled       | true           | boolean | Set to true to split non-indexed queries by shards splits. Should generally be
| es6.scroll_search_batch_size            | 100            | integer | max of 100 hits will be returned for each scroll. Default 100                 
| es6.scroll_search_timeout               | 60000          | bigint  | If set, will enable scrolling of the search request for the specified timeout

demo:```set session es6.scroll_search_batch_size=150```

## Does not support features
+ insert into(Already supported)
+ create table as (Already supported)
+ create table (Already supported)
+ drop table(Already supported)
+ update
+ delete

### CODE REVIEW

**Comment by harbby** [comment]
@nezihyigitbasi  
Thank you very much for your attention, I will add a detailed instructions for use.
I think this program is very interesting.
I implemented a very cool predicate pushdown method that can very well leverage the search capabilities of elasticsearch.
This method can also be used in the mogodb connector.
In addition, she also supports the latest elasticsearch5 and elasticsearch6

**Comment by harbby** [comment]
@kokosing  
Hello sir
Here is a PR about the elasticsearch connector, can you follow it?
thank you very much

File: elasticsearch-apache2/pom.xml [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+                    <execution>
+                        <id>default</id>
+                        <phase>verify</phase>
+                        <goals>
+                            <goal>check</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <!-- 在发现重复java类的情况下,是否构建失败  :设置flase为构建不失败 true为构建失败 -->
```
Comment (kokosing) [inline elasticsearch-apache2/pom.xml]: Please use english in every comment

File: elasticsearch-apache2/pom.xml [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+            <version>${dep.elasticsearch.version}</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>junit</groupId>
+                    <artifactId>junit</artifactId>
+                </exclusion>
+                <!--<exclusion>-->
+                <!--<artifactId>joda-time</artifactId>-->
+                <!--<groupId>joda-time</groupId>-->
+                <!--</exclusion>-->
```
Comment (kokosing) [inline elasticsearch-apache2/pom.xml]: remove

File: elasticsearch-apache2/pom.xml [hunk 3]
```diff
@@ -0,0 +1,387 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kokosing) [inline elasticsearch-apache2/pom.xml]: Why shading is needed?

File: presto-base-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchConnector.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            ElasticsearchPageSinkProvider pageSinkProvider,
+            ElasticsearchSessionProperties sessionProperties)
+    {
+        this.connectorId = requireNonNull(connectorId, "connectorId is null");
+        this.client = requireNonNull(client, "client is null");
+        this.splitManager = requireNonNull(splitManager, "splitManager is null");
+        this.pageSourceProvider = requireNonNull(pageSourceProvider, "pageSourceProvider is null");
+        this.pageSinkProvider = requireNonNull(pageSinkProvider, "pageSinkProvider is null");
+        this.sessionProperties = requireNonNull(sessionProperties, "sessionProperties is null");
+//        this.tableProperties = requireNonNull(tableProperties, "tableProperties is null");
```
Comment (kokosing) [inline presto-base-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchConnector.java]: remove such comments

File: presto-base-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchConnector.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    private static final Logger LOG = Logger.get(ElasticsearchConnector.class);
+    private final ConcurrentMap<ConnectorTransactionHandle, ElasticsearchMetadata> transactions = new ConcurrentHashMap<>();
+
+    private final ElasticsearchConnectorId connectorId;
+    private final BaseClient client;
+    private final ElasticsearchSplitManager splitManager;
+    private final ElasticsearchPageSourceProvider pageSourceProvider;
+    private final ElasticsearchPageSinkProvider pageSinkProvider;
+    private final ElasticsearchSessionProperties sessionProperties;
+//    private final ElasticsearchTableProperties tableProperties;
```
Comment (kokosing) [inline presto-base-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchConnector.java]: remove

File: presto-elasticsearch6/src/main/java/com/facebook/presto/elasticsearch6/functions/MatchQueryFunction.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import com.facebook.presto.spi.function.SqlNullable;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+
+/**
+ * 输入bitmap 输出该bitmap的值
```
Comment (kokosing) [inline presto-elasticsearch6/src/main/java/com/facebook/presto/elasticsearch6/functions/MatchQueryFunction.java]: Use english please

**Comment by findepi** [comment]
@harbby @kokosing before you continue with review here, i think it would be worthy to get some follow-up to my comment in https://github.com/prestodb/presto/pull/9459#issuecomment-404521131. Citing here for convenience:
> @harbby can you coordinate with @zhenxiao so that we can get the best of https://github.com/prestodb/presto/pull/10999 and this (https://github.com/prestodb/presto/pull/9459) ?

File: presto-elasticsearch6/src/main/java/com/facebook/presto/elasticsearch6/functions/MatchQueryFunction.java (hunk 6 above)
Comment (harbby) [reply presto-elasticsearch6/src/main/java/com/facebook/presto/elasticsearch6/functions/MatchQueryFunction.java]: @kokosing Sorry, ok, I will modify it as soon as possible.

**Comment by zhenxiao** [comment]
thank you @findepi 

Hi @harbby @kokosing shall we sync for Presto Elasticsearch Connector work?

Here is the PR:
https://github.com/prestodb/presto/pull/9459

The PR has been running at Uber in production for 6+ months, we also have support for json functions pushdown, limit pushdown, predicate pushdown stuff, will submit as separate PRs, as they are mostly in optimizers.

We also have support for ES2, ES5, and ES6, and are running production use cases. submit ES5 PR for review is a first step.

Our current status is working on integration tests, almost finished, will send an update soon.

keep @nezihyigitbasi posted

**Comment by kokosing** [comment]
Can we close this in favor of #9459 ?

**Comment by harbby** [comment]
@kokosing  Yes

**Pull request closed without merging**